lib/Pipeline/Step.cpp

```diff
@@ -144,40 +330,144 @@ void Step::removeSatisfiedGoals(ContainerToTargetsMap &Targets,
                                 ContainerToTargetsMap &ToLoad) const {
   for (auto &RequiredInputsFromContainer : Targets) {
     llvm::StringRef ContainerName = RequiredInputsFromContainer.first();
-    auto &RequiredInputs = RequiredInputsFromContainer.second;
-    auto &ToLoadFromCurrentContainer = ToLoad[ContainerName];
+    TargetsList &RequiredInputs = RequiredInputsFromContainer.second;
+    TargetsList &ToLoadFromCurrentContainer = ToLoad[ContainerName];
     if (Containers.contains(ContainerName))
       removeSatisfiedGoals(RequiredInputs,
                            Containers.at(ContainerName),
                            ToLoadFromCurrentContainer);
   }
 }
 
-ContainerToTargetsMap Step::deduceResults(const Context &Ctx,
-                                          ContainerToTargetsMap Input) const {
-  for (const auto &Pipe : Pipes)
-    Input = Pipe->deduceResults(Ctx, Input);
+ContainerToTargetsMap Step::deduceResults(ContainerToTargetsMap Input) const {
+  for (const PipeWrapper &Pipe : Pipes)
+    Input = Pipe.Pipe->deduceResults(*Ctx, Input);
   return Input;
 }
 
 Error Step::invalidate(const ContainerToTargetsMap &ToRemove) {
+  for (auto &Pipe : Pipes) {
+    Pipe.InvalidationMetadata.remove(ToRemove);
+  }
   return containers().remove(ToRemove);
 }
 
 Error Step::store(const revng::DirectoryPath &DirPath) const {
-  return Containers.store(DirPath);
+  if (auto Error = Containers.store(DirPath))
+    return Error;
+
+  return storeCacheToDisk(DirPath);
 }
 
-Error Step::checkPrecondition(const Context &Ctx) const {
-  for (const auto &Pipe : Pipes) {
-    if (auto Error = Pipe->checkPrecondition(Ctx); Error)
+Error Step::checkPrecondition() const {
+  for (const PipeWrapper &Pipe : Pipes) {
+    if (llvm::Error Error = Pipe.Pipe->checkPrecondition(*Ctx); Error)
       return llvm::make_error<AnnotatedError>(std::move(Error),
                                               "While scheduling pipe "
-                                                + Pipe->getName() + ":");
+                                                + Pipe.Pipe->getName() + ":");
   }
   return llvm::Error::success();
 }
 
 Error Step::load(const revng::DirectoryPath &DirPath) {
-  return Containers.load(DirPath);
+  auto MaybeBool = DirPath.exists();
+  if (not MaybeBool)
+    return MaybeBool.takeError();
+
+  if (not MaybeBool.get())
+    return llvm::Error::success();
+
+  if (auto Error = Containers.load(DirPath))
+    return Error;
+
+  return loadCacheFromDisk(DirPath);
```

Comment:
Feels like `loadCacheFromDisk` should be squashed into `Containers.load`, unless there's a reason to split them.

Reply:
i did it this way just because it easier to read.